---
title: 'Microbiome Analysis: analysis preamble `r emo::ji("person_climbing")`'
author: 'jfg'
date: "`r format(Sys.time(), '%d %b %Y, %H:%M')`"
output:
  html_document:
    toc: TRUE
    toc_depth: 3
    toc_float: TRUE
    number_sections: FALSE
  pdf_document: 
    toc: TRUE
    toc_depth: 3
    number_sections: FALSE
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, encoding=encoding, output_file='../../documents/mb6302__preamble.html') })
---


---


# Preamble

---
    
#### what this **is**, and **who** it is for:

This document covers how to do the following:

  0. connect through `ssh` and set up a `bash`-based bioinformatics project
  1. collect and check a project's illumina short-read dataset (QC - `FastQC+MultiQC`)
  2. improve the overall quality of their sequencing dataset (QC - `trimmomatic`)
  3. remove contaminant and/or host sequences from the dataset (QC - `kneaddata`)
  4. generate a simple but effective community profile of the microbial community in that dataset (Microbial ecology - `Kraken2+Bracken`)

We provide _minimal_ guidance on how to adapt the different steps to the workflow (different parameters, different options or tools). 

This document was originally put together by Ines Calvete (IPLA/CSIC), Nuria Pe«πa Vidal (UCM), and Jamie FitzGerald (Teagasc/UCD) as a roadmap for analysing microbiomes from cheese, and chicken GI tract respectively. It uses material sourced from well-written reference manuals, online communities, and our own mistakes (see *References* section). We assume a beginner's understanding of the HPC, `ssh`, `bash`, and scripting.  

Please note this page is still being written. If something is missing, unclear, or flaming wrong, please point it out! 
<!-- See the **Intro document which is not written yet**.-->

---

#### what this document is **not**:

<!-- , and the intro document (**Nov 2022: not yet written**). -->
This is not a guide to the _**Why**_ of the methods used here. For that, check the *References* section at the bottom of the page. In particular, importing data, and analysis of communities, in `R` will need their own documents:

  * [importing data to `R` `r emo::ji("computer")`](./data_to_R.html)
  
  * analysing microbial communities in R: this is a big topic - consider the links below (and bottom of page):
  
    - [Data Carpentry's Microbial Ecology Workshop](https://cnoecker.github.io/R-microbiome-workshop/index.html)
    - [AstroBioMike - Bioinformatics for beginners](https://astrobiomike.github.io/)
    - [microbiome code browser](http://metagenome.cs.umn.edu/microbiomecodebrowser)

---


#### when something goes **wrong**:

Mistakes, and trial-and-error, are part of learning any skill. Bioinformatics involves a particularly large and constant amount of trialing, as well as lots of errors (see: [**how to read an error**](https://www.baeldung.com/linux/bash-errors)).

One of the best ways to learn quickly (and hopefully make different, exciting mistakes rather than the same ones) is to keep notes of what you do, just like in a lab:

  * write out your objectives
  * date your notes
  * write out what you try to do
  * note down the outcomes
  * write down what goes wrong
  * include the images if you can
  
---


#### if you need **help**:

  * in `bash`: for some e.g. `program`, try entering `program --help`, `program -h`, or `man program`..
  * in `R`: for a `function()`, enter `?function` to see that function's help page, or try pressing `tab` after a comma to see other options that `function()` understands.
  * in `R`: for some e.g. `object` (dataframe, matrix, list, etc.) you can try using the `str(object)` function to see its structure, use `head(object)` to see the top 6 lines or rows, `class(object)` to see what sort of thing it is, `dim(object)` to see it's dimensions (`row x columns`; this will be null if it's a list or vector etc., as these only have 1 row and 0 columns, and `1*0 = 0`)
  * try searching for your `error message` online (web search, stackoverflow, biostars, bioconductor, github, huttenhower-biobakery, etc.) - the issue might be simple! If that doesn't work, or you're not sure: 
  * _**Ask someone**_. This is not easy work, and we all know it. People are glad to help, and to show others what they have found out.

---


> Finally, please remember:

>  * when you login to the HPC, you must **always** change to a `node` (eg `ssh compute05`) before you do any work 


# Microbiome Analysis

  * <a href="mb6302__setup.html">`workspace setup`</a>
  * <a href="mb6302__qc.html">`sequence QC`</a>
  * <a href="mb6302__decontam.html">`sequence decontamination`</a>
  * <a href="mb6302__taxonomy.html">`sequence taxonomic estimation`</a>




---

# Reading / Reference

### - linux `101`

> [Linux command line for beginners](https://ubuntu.com/tutorials/command-line-for-beginners) - quite quite good!

> [`slurm` and variables](https://help.rc.ufl.edu/doc/Using_Variables_in_SLURM_Jobs)

> [more `slurm` formatting](https://confluence.ecmwf.int/display/UDOC/Writing+SLURM+jobs)

> [`glomming`](https://linuxhint.com/bash_globbing_tutorial/)

> [regular expressions (`regex`)](https://robinwinslow.uk/regex)

> [`if-statements - controlling what happens](https://ryanstutorials.net/bash-scripting-tutorial/bash-if-statements.php)

> [`for-loop`s - first step in `bash` superpowers](https://linuxize.com/post/bash-for-loop/)

> [How to read an error](https://www.baeldung.com/linux/bash-errors)

> [info on variables](https://ostechnix.com/bash-variables-shell-scripting/)


### - DNA sequencing & co

> [`FastQC`](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) 

> [`MultiQC` site + support](https://multiqc.info)

> [`Trimmomatic` site](http://www.usadellab.org/cms/index.php?page=trimmomatic)

> [`Trimmomatic` manual](http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf)

> [`bowtie2 / samtools` filtering](https://www.metagenomics.wiki/tools/short-read/remove-host-sequences)


### - metagenomics 

> [AstroBioMike - Bioinformatics for beginners](https://astrobiomike.github.io/) - extremely good and thorough guide.

> [The `Kraken` **Suite** ](http://ccb.jhu.edu/data/kraken2_protocol/) - **Sept 2022** nature [paper](https://doi.org/10.1038/s41596-022-00738-y) from the developers of Kraken, KrakenUniq, & Bracken etc. 

> [microbiome code browser](http://metagenome.cs.umn.edu/microbiomecodebrowser) - look v good also - Dan Knights & co., 2016
