---
title: 'shotgun metagenomics - microbial communities - estimating composition'
author: 'IC / NPV / JFG'
date: "`r format(Sys.time(), '%d %b %Y, %H:%M')`"
output:
  html_document:
    toc: TRUE
  pdf_document: 
    toc: TRUE
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, encoding=encoding, output_file='../documents/4.microcomm.html') })
---


`article in suspension`


<!-- # 4 - microbial community profiling -->

<!-- The dataset is as clean and as well-filtered as can be expected. The next step is (can be) extrapolating the composition of the microbial community from the sequence data corresponding to the different samples. To add additional emphasis: it is not possible to perfectly recapture the microbiome of an environment: this is a ~random sample of a community; biases are introduced at all steps; databases are necessarily incomplete; sequencing is an imperfect technology; many sequences are non-informative; exhaustive identification is generally infeasible, etc.   -->

<!-- Nevertheless, we are very interested in _estimating_ the community composition as best we can. [`Kraken2`]() is a popular method of assigning taxonomic identities to individual reads (if possible) based on the idea that different organisms naturally exhibit different base-preferences, so will have a characteristic probability for any given sequence that is `$k`bases long. At a given value of $k$, organisms can be reliably distinguished by the abundance of different length-$k$ [oligomers](https://en.wikipedia.org/wiki/Oligomer) (a.k.a. `k-mers`). `Kraken2` is one of a number of programs which use variations on this approach, and is popular because of it continued development of both software and databases, its continued support by the authors, relative ease of use, and the availbility of it's companion program, [`Bracken`]().  -->

<!-- Two issues with `Kraken` are that it will not force e.g. a species identity on a given read if it cannot be reliably assigned; and that the "count" of reads assigned to a particular tip of the tree (e.g. species or strain) cannot reliably be taken as the actual `abundance` of that species/strain in that sample: it is instead a count of the number of reads that were unique enough to be classified to a taxonomic identity. -->

<!-- This lead to many reads being "stranded" across the phylogenetic tree, reducing the dataset size further. Bracken solves this "stranding" by regrouping all sequences to a user-defined taxonomic level (e.g. `species`), and uses a Bayesian re-estimate of the abundances to provide a measure of the likely proportional abundance of each taxon in the dataset.  -->

<!-- This might seem like an unnecessary complication, but it has been shown that this Bayesian re-estimation of microbial composition significantly improves the measures generated of microbial communities. It's also super fast - nearly instant - so why would you not? -->


<!-- ### Using Kraken 2 -->

<!-- ```{bash, eval = FALSE} -->

<!-- usearch -fastq_join $KDOUT/$SAMPLE/${SAMPLE}_R1_kneaddata_paired_R1.fastq.gz -reverse $KDOUT/$SAMPLE/${SAMPLE}_R1_paired_R2.fastq.gz -fastaout $KDOUT/${SAMPLE}_kneaddata_R1R2.fastq.gz -->

<!-- # --------- -->

<!-- # test -->
<!-- K2SAMPLE=SC145C1-NS_S86 -->
<!-- K2IN=$KDOUT -->
<!-- K2OUT=$KROUT -->
<!-- K2REFDIR=/data/databases/Kraken2_GTDB_r89_54k -->

<!-- # recall - first run requires a load to shared mem step, which takes time  -  to do this, possibly omit the memory mapping step. GOing to omit entirely as mem not an issue, yet -->
<!-- cat $K2IN/${K2SAMPLE}/${K2SAMPLE}_R1_kneaddata_Homo*.fastq.gz > $K2IN/${K2SAMPLE}/${K2SAMPLE}_contam.fastq.gz & -->
<!-- cat $K2IN/${K2SAMPLE}/${K2SAMPLE}_R1_kneaddata_unmatched_[12].fastq.gz > $K2IN/${K2SAMPLE}/${K2SAMPLE}_orphan.fastq.gz & -->
<!-- # THREE MINUTES! only 18% classified though.  | 0.5 conf: 2m, 2% class | 0.15 cutoff - 3m, 13% ID'd -->
<!-- time kraken2 --db $K2REFDIR  $K2IN/${K2SAMPLE}/${K2SAMPLE}_R1_kneaddata_paired_1.fastq.gz $K2IN/${K2SAMPLE}/${K2SAMPLE}_R1_kneaddata_paired_2.fastq.gz --paired --confidence 0.15 --gzip-compressed --threads 5 --report-zero-counts --report $K2OUT/${K2SAMPLE}_test_kraken_report# --unclassified-out $K2OUT/${K2SAMPLE}_test_kraken_unclass# --output $K2OUT/${K2SAMPLE}_test_kraken_output# -->
<!-- # TWO-POINT-FIVE MINUTES! 27% identified  | conf 0.5: 1m30, 2% identifiedi | 3 MIN 30 - 15% classified -->
<!-- time kraken2 --db $K2REFDIR  $K2IN/${K2SAMPLE}/${K2SAMPLE}_orphan.fastq.gz --gzip-compressed --confidence 0.15 --threads 5 --report-zero-counts --report $K2OUT/${K2SAMPLE}_test_orphan_kraken_report --unclassified-out $K2OUT/${K2SAMPLE}_test_orphan_kraken_unclass --output $K2OUT/${K2SAMPLE}_test_orphan_kraken_output -->
<!-- # THREE MIN FORTY! 12% sequences identified, highly worrying. -->
<!-- time kraken2 --db $K2REFDIR  $K2IN/${K2SAMPLE}/${K2SAMPLE}_contam.fastq.gz --gzip-compressed --confidence 0.15 --threads 5 --report-zero-counts --report $K2OUT/${K2SAMPLE}_test_contam_kraken_report --unclassified-out $K2OUT/${K2SAMPLE}_test_contam_kraken_unclass --output $K2OUT/${K2SAMPLE}_test_contam_kraken_output -->

<!-- # need to specify id cutoff -->

<!-- # for ROUND 1, paired-mode only is fine, as some disagreement about handling btw forum and manual - possibly has been updated in latter. -->
<!-- # can remember why we included the zero counts...but we did, intentionally -->

<!-- echo '#!/bin/bash -->

<!-- #SBATCH –-job-name=knead_fastq -->
<!-- #SBATCH –-output=knead_fastq.txt -->

<!-- #SBATCH –-ntasks=5 -->
<!-- #SBATCH –-time=40:00 -->

<!-- SAMPLE=$1 -->
<!-- IN=$2 -->
<!-- OUT=$3 -->
<!-- REFDIR=$4 -->

<!-- time kraken2 --db $REFDIR  $K2IN/${K2SAMPLE}/${K2SAMPLE}_R1_kneaddata_paired_1.fastq.gz $K2IN/${K2SAMPLE}/${K2SAMPLE}_R1_kneaddata_paired_2.fastq.gz --paired --confidence 0.15 --gzip-compressed --threads 5 --report-zero-counts --use-mpa-style --report $K2OUT/${K2SAMPLE}_kraken_report# --unclassified-out $K2OUT/${K2SAMPLE}_kraken_unclass# --output $K2OUT/${K2SAMPLE}_kraken_output# -->

<!-- ## -->
<!-- ## here, should cat and grab the orphan output -->
<!-- ## -->

<!-- scp $OUT/${SAMPLE}* jamie@143.239.204.169:/mnt/yabba/yoda_backup/home/ClaessonFTP/jamie/ -->

<!-- ## within dir, do fastqc -o $QCDIR/ -t 4 $OUT/${SAMPLE}/*.fastq.gz -->


<!-- ' > $MAT/slurm_krak2.sh -->

<!-- ls $FILT/*R[12].fastq.gz | sed -e 's/.*\/\(.*\)_R..*/\1/g' | sort | uniq | head -20 | tail -1 | parallel -j 8 sbatch $MAT/slurms/slurm_krak2.sh {} $KDOUT $KROUT $K2REFDIR -->
<!-- ls $FILT/*R[12].fastq.gz | sed -e 's/.*\/\(.*\)_R..*/\1/g' | sort | uniq | head -20 | tail -1 | parallel -j 8 $MAT/slurms/bash_krak2.sh {} $KDOUT $KROUT $K2REFDIR -->


<!-- # lifeboat protocol -->
<!-- # K2SAMPLE=SC145C1-NS_S86 -->
<!-- K2IN=/claesson/jamie/fhi_redch -->
<!-- K2OUT=/claesson/jamie/fhi_redch/krak2 -->
<!-- K2REFDIR=/claesson/jamie/ref/kraken2_stnd -->


<!-- # << < < < need to add step here to start with one sample, removing --mem-map, and updating dirs/ -->
<!-- ## remove --mpa-style as interferes with Bracken -->
<!-- ls /claesson/jamie/fhi_redch/*[12].fastq.gz | sed -e 's/.*\/\(.*\)_R..*/\1/g' | sort | uniq | parallel -j 6 time kraken2 --db $K2REFDIR  $K2IN/{}_R1_kneaddata_paired_1.fastq.gz $K2IN/{}_R1_kneaddata_paired_2.fastq.gz --memory-mapping --paired --confidence 0.15 --gzip-compressed --threads 5 --report-zero-counts --report $K2OUT/{}_kraken_report# --unclassified-out $K2OUT/{}_kraken_unclass# --output $K2OUT/{}_kraken_output# >> ~/para_krak_fhi_redch.stout 2>&1 -->

<!-- # this'll put out all the seq too. if you only want the report, send your output/unclassified ..away -->
<!-- # try catch output, or anything else - because where does it go? -->

<!-- ## pre-run very possibly not necessary with --memorymapping opt -->



<!-- ##   B R A C K E N   B E A T I N G       ## =================== -->

<!-- brthreads=10 -->
<!-- b=$DAN/3_b2  ; mkdir $b -->
<!-- binst=~/BioApps -->

<!-- # K2SAMPLE=SC145C1-NS_S86 -->
<!-- WRK=/claesson/jamie/fhi_redch -->
<!-- K2IN=/claesson/jamie/fhi_redch/knead -->
<!-- K2OUT=/claesson/jamie/fhi_redch/krak2 -->
<!-- K2REFDIR=/claesson/jamie/ref/kraken2_stnd -->

<!-- BRK=$WRK/brack ; mkdir $BRK -->

<!-- ## build data base first, using length 80bp 35kmers, takes an hour on 25 threads -->

<!-- ## fast as immediate  -  run bracken on the kraken_report  -  note no threads arg! -->
<!-- # -r = read lengths (taken as the starting read length, we're at about 125 by now...) -->
<!-- # -l = Level, def=S ; (K, P, C, O, F, G, S ) -->
<!-- # -t = threshold, def = 10 ; level below which a taxo-branch on the kraken tree will be dropped from further assignment of abundances -->
<!-- ls $K2OUT/* | sed -r 's/.*\/(.*)_kraken.*/\1/g' | sort | uniq | \ -->
<!-- parallel -j 3  \ -->
<!-- python ~/BioApps/Bracken/src/est_abundance.py -i $K2OUT/*{}*report -k $K2REFDIR/database80mers.kmer_distrib -l S -t 10 -o $BRK/{}.bracken \ -->
<!-- > $BRK/bracken_output_82.stout -->
<!-- # bracken -d $DB/kraken2_stnd -i $KRK/*{}_*report -o $b/{}_bracken_report.txt -r 150 -l S -t 10 > $b/bracken.stout -->

<!-- ## combine reports to one monolith.tsv -->
<!-- python2 ~/BioApps/Bracken/analysis_scripts/combine_bracken_outputs.py --files $BRK/*bracken -o fhi_redch_krakenStnd_bracken_combo -->

<!-- ## convert to MPA-style report outputs, in order to get taxonomic tree (again, surely a better way than this) -->
<!-- parallel python2 ~/BioApps/Bracken/src/kreport2mpa.py -r {} -o $BRK/{.}.mpa ::: $K2OUT/*report* > $BRK/fhi_redch_bragglers -->


<!-- # # catch done/undone -->
<!-- #   find $KOUT/*gz | sed -r 's/.*[\/_]([0-9]{4}[AB]?)[\/_].*/\1/g' | sort | uniq > $DAN/Materials/temp2 -->
<!-- #   find $KRK/*report | sed -r 's/.*[\/_]([0-9]{4}[AB]?)[\/_].*/\1/g' | sort | uniq > $DAN/Materials/temp3_k -->
<!-- #   comm -23 $DAN/Materials/temp2 $DAN/Materials/temp3_k > $DAN/Materials/krk_offenders.txt -->
<!-- #   # include switch for outputting zero-reads - for tabulation -->
<!-- # -->

<!-- ##   T A X O N O M Y   -   glom from mpa output!   ## ======================================== -->

<!-- ## own funciton gettax_kraken : hacky, ++slow, and only gets species name. Avoid. -->

<!-- ## get all the S level assignments; trim counts ; sort; uniq -->
<!-- grep -h '|s_' $BRK/*mpa | cut -f 1 > $BRK/bracken_82_taxa.temp -->
<!-- sort $BRK/bracken_82_taxa.temp | uniq > $BRK/bracken_82_taxa.temp2 -->
<!-- sed 's/|/\t/g' $BRK/bracken_82_taxa.temp2 > $BRK/fhi_redch_krakenStnd_bracken_combo_taxa_82 -->

<!-- ## send home -->
<!-- scp ./bracken_545_*[combo,taxa] $DSK -->

<!-- # consider total taxonomy, but prob doesn't tell us anything re: missing levels -->
<!-- cat $b/bracken_reports_545/*mpa | cut -f 1 > $b/bracken_545_total_tax.temp -->
<!-- sort  $b/bracken_545_total_tax.temp | uniq > $b/bracken_545_total_tax.temp2 -->


<!-- ``` -->


<!-- # Checking your output -->
<!-- ```{bash, eval = FALSE} -->
<!-- ``` -->


<!-- --- -->

<!-- # this document isn't finished! -->

<!-- This document is still being written. It still needs the following steps: -->

<!--   * not really explaining the kneaddata part properly -->
<!--   * feed clean sequences to `Kraken2` -->
<!--   * sanitise the `Kraken2` output using `Bracken` -->
<!--   * start doing the microbial ecology in `R`! But that's a story for another day... -->

<!-- --- -->

<!-- # Reading / Reference -->

<!-- > kneaddata -->

<!-- > genome data... -->

<!-- > decoy genome -->

<!-- > bowtie2... -->